numba/tests/test_np_functions.py

```diff
@@ -2175,6 +2183,80 @@ def a_variations():
 
         self.assertIn("Cannot np.flip on UniTuple", str(raises.exception))
 
+    def _check_split(self, func):
+        # Since np.split and np.array_split are very similar
+        pyfunc = func
+        cfunc = jit(nopython=True)(pyfunc)
+
+        def args_variations():
+            a = np.arange(100)
+            yield a, 2
+            yield a, 2, 0
+            yield a, [1, 4, 72]
+            yield list(a), [1, 4, 72]
+            yield tuple(a), [1, 4, 72]
+            yield a, [1, 4, 72], 0
+            yield list(a), [1, 4, 72], 0
+            yield tuple(a), [1, 4, 72], 0
+
+            a = np.arange(64).reshape(4, 4, 4)
+            yield a, 2
+            yield a, 2, 0
+            yield a, 2, 1
+            yield a, [2, 1, 5]
+            yield a, [2, 1, 5], 1
+            yield a, [2, 1, 5], 2
+            yield a, [1, 3]
+            yield a, [1, 3], 1
+            yield a, [1, 3], 2
+
+            a = np.arange(100).reshape(2, -1)
+            yield a, 1
+            yield a, 1, 0
+            yield a, [1], 0
+            yield a, 50, 1
+            yield a, np.arange(10, 50, 10), 1
+            yield a, (1,)
+            yield a, (np.int32(4), 10)
```

Comment:
I added these locally:
```
            a = np.array([])
            yield a, 1
            yield a, 2
            yield a, (2, 3), 0
            yield a, 1, 0

            a = np.array([[]])
            yield a, 1
            yield a, (2, 3), 1
            yield a, 1, 0
            yield a, 1, 1
```
a few fail. Having looked at the NumPy code, I am wondering if the behaviour is a side effect of the implementation opposed to a specific design. Am considering whether this sort of thing is intended:
```
In [51]: np.array_split(np.array([[[]]]), 3)
Out[51]: 
[array([], shape=(1, 1, 0), dtype=float64),
 array([], shape=(0, 1, 0), dtype=float64),
 array([], shape=(0, 1, 0), dtype=float64)]
```

Reply:
The shapes of those arrays are pretty weird though.

Which fail for you? I'm only seeing `split` tests failing (not `array_split`), and it looks like the problem there is that either version would error.

Reply:
Seems like these fail:
```
--------------------------------------------------------------------------------
(array([], dtype=float64), 1)
NB Maximum allowed size exceeded
--------------------------------------------------------------------------------
(array([], dtype=float64), 2)
NB Maximum allowed size exceeded
--------------------------------------------------------------------------------
(array([], dtype=float64), 1, 0)
NB Maximum allowed size exceeded
--------------------------------------------------------------------------------
(array([], shape=(1, 0), dtype=float64), 1, 1)
NB Maximum allowed size exceeded
```
under `np.split`, I think the issue is here:
https://github.com/numba/numba/pull/5699/files#diff-ef28bfb9dfcf3597915208b3329a6ac6R5046
specifically:
```
np.arange(l, ary.shape[axis], l),
```
if `l` is zero then it leads to a zero step size in the `arange` which is where that error message is coming from.

> The shapes of those arrays are pretty weird though.

yeah, this is why I'm thinking it might be a side effect opposed to intended? What do you think?

Reply:
Ah, found it. I was doing something wrong with `split` and have added these test cases. It's possible the implementation could be cleaner if I just passed the input integer after the check (leading to an additional `divmod` call). 

Is the compiler smart enough to skip the additional `divmod` call if that approach were taken?

Reply:
> Ah, found it. I was doing something wrong with `split` and have added these test cases. It's possible the implementation could be cleaner if I just passed the input integer after the check (leading to an additional `divmod` call).

Great, thanks.

> Is the compiler smart enough to skip the additional `divmod` call if that approach were taken?

Can you show me what you mean please? Where would the `divmod` end up? I can probably make a guess from there :)

Reply:
> Can you show me what you mean please? Where would the divmod end up? I can probably make a guess from there :)

Ah, yes, I commented before I realized the tests hadn't finished running. Here's what I'm talking about: https://github.com/numba/numba/pull/5699/commits/93a9218f01c79ca00ea8f0b906f795bb3e2887d2#r500160142

> yeah, this is why I'm thinking it might be a side effect opposed to intended?

I think this might be like passing indices that aren't monotonic. I can't imagine a case where not throwing an error is what someone wants, but it is the behavior numpy has. I've got that behavior replicating now though.

Reply:
Great, thank you.